executor/new_join.go

```diff
@@ -0,0 +1,248 @@
+		e.outerRows = e.outerRows[:0]
+		e.resultRows = e.resultRows[:0]
+		e.innerDatums = e.innerDatums[:0]
+		for i := 0; i < batchSize; i++ {
+			outerRow, err := e.children[0].Next()
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if outerRow == nil {
+				e.exhausted = true
+				break
+			}
+			match, err := expression.EvalBool(e.leftConditions, outerRow.Data, e.ctx)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if match {
+				joinDatums := make([]types.Datum, 0, len(e.outerJoinKeys))
+				for _, col := range e.outerJoinKeys {
+					datum, _ := col.Eval(outerRow.Data)
+					joinDatums = append(joinDatums, datum)
+				}
+				joinOuterEncodeKey, err := codec.EncodeValue(nil, joinDatums...)
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				e.outerRows = append(e.outerRows, orderRow{key: joinOuterEncodeKey, row: outerRow})
+				e.innerDatums = append(e.innerDatums, orderRow{key: joinOuterEncodeKey, row: &Row{Data: joinDatums}})
+			} else if e.outer {
+				e.resultRows = append(e.resultRows, e.fillDefaultValues(outerRow))
+			}
+		}
+		sort.Sort(e.outerRows)
+		err := e.doJoin()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		e.cursor = 0
+	}
+	row := e.resultRows[e.cursor]
+	e.cursor++
+	return row, nil
+}
+
+func (e *IndexLookUpJoin) fillDefaultValues(row *Row) *Row {
+	row.Data = append(row.Data, e.defaultValues...)
+	return row
+}
+
+func uniqueOrderedRows(rows orderedRows) [][]types.Datum {
+	datums := make([][]types.Datum, 0, rows.Len())
+	sort.Sort(rows)
+	for i := range rows {
+		if i > 0 && bytes.Compare(rows[i-1].key, rows[i].key) == 0 {
+			continue
+		}
+		datums = append(datums, rows[i].row.Data)
+	}
+	return datums
+}
+
+func (e *IndexLookUpJoin) doJoin() error {
+	err := e.innerExec.doRequestForDatums(uniqueOrderedRows(e.innerDatums), e.ctx.GoCtx())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer e.innerExec.Close()
+	for {
+		innerRow, err := e.innerExec.Next()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if innerRow == nil {
+			break
+		}
+		match, err := expression.EvalBool(e.rightConditions, innerRow.Data, e.ctx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if !match {
+			continue
+		}
+		joinDatums := make([]types.Datum, 0, len(e.innerJoinKeys))
+		for _, col := range e.innerJoinKeys {
+			datum, _ := col.Eval(innerRow.Data)
+			joinDatums = append(joinDatums, datum)
+		}
+		joinKey, err := codec.EncodeValue(nil, joinDatums...)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		e.innerRows = append(e.innerRows, orderRow{key: joinKey, row: innerRow})
+	}
+	sort.Sort(e.innerRows)
+	return e.doMergeJoin()
+}
+
+func getNextCursor(cursor int, rows orderedRows) int {
+	for {
+		cursor++
+		if cursor >= len(rows) {
+			break
+		}
+		c := bytes.Compare(rows[cursor].key, rows[cursor-1].key)
+		if c != 0 {
+			break
+		}
+	}
+	return cursor
+}
+
+func (e *IndexLookUpJoin) doMergeJoin() error {
```

Comment:
Add a few comments for this function. Such as:
e.innerRows and e.outerRows are already sorted, so we can do SortMergeJoin on the two data source.